### PR TITLE
init: Remove auto-import of `bootstrap.dat` and associated code

### PR DESCRIPTION
**Pull request by fjahr**
This picks up #15954 

I fixed the code and added at a functional test utilizing the scripts in `contrib/linearize` as suggested by @MarcoFalke .

### CODE REVIEW

File: test/functional/feature_loadblock.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python3
+# Copyright (c) 2017-2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test loadblock option
+
+Test the option to start a node with the option loadpath which
```
Comment (mzumsande) [inline test/functional/feature_loadblock.py]: loadblock, not loadpath

File: test/functional/feature_loadblock.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+loads a serialized blockchain from a file (usually called bootstrap.dat).
+To generate that file this test uses the helper scripts in available
```
Comment (mzumsande) [inline test/functional/feature_loadblock.py]: remove "in"

**Review by mzumsande — COMMENTED** [review]
The CIs don't like the new test yet, probably because the relative paths require one to be in the bitcoin root dir for the scripts to be found.

Also, the help text for `loadblock`, "Imports blocks from external blk000??.dat file on startup" could be updated to remove the file name.

File: test/functional/feature_loadblock.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+hashlist={hashlist}""".format(datadir=data_dir,
+                              output_file=bootstrap_file,
+                              user=node_url.username,
+                              password=node_url.password,
+                              port=node_url.port,
+                              host=node_url.hostname,
+                              hashlist=hashlist_path,
+                              block_dir=block_dir,
+                              genesis=genesis_block))
+        cfg.close()
```
Comment (MarcoFalke) [inline test/functional/feature_loadblock.py]: Can use https://docs.python.org/3/reference/compound_stmts.html#with instead?

And the writing it like this, should make the test shorter in line count. https://github.com/bitcoin/bitcoin/blob/7b701fef58f627956d597817a1f9422edd890cdc/test/functional/test_framework/util.py#L306

Comment (fjahr) [reply test/functional/feature_loadblock.py]: Yes, that's much better, implemented that.

File: test/functional/feature_loadblock.py (hunk 2 above)
Comment (fjahr) [reply test/functional/feature_loadblock.py]: fixed

File: test/functional/feature_loadblock.py (hunk 1 above)
Comment (fjahr) [reply test/functional/feature_loadblock.py]: fixed

File: test/functional/feature_loadblock.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        os.system(linearize_data_path + " " + cfg_path)
+
+        self.log.info("Restart node with bootstrap file")
+        self.stop_node(0)
+
+        # use a new empty blocks dir to make sure blocks are loaded from bootstrap file
+        shutil.rmtree(self.nodes[0].datadir)
+        initialize_datadir(self.options.tmpdir, 0, self.chain)
+        blocksdir_path = os.path.join(self.options.tmpdir, 'blocksdir')
+        os.mkdir(blocksdir_path)
```
Comment (MarcoFalke) [inline test/functional/feature_loadblock.py]: What is the point of `blocksdir_path` (new empty blocks dir), when the whole datadir is cleaned?

I'd even go further and not remove the whole datadir tree. Since that also deletes the debug log for debugging purposes.

You could spin up two nodes, and call `setnetworkactive(False)` on the second one. Then reboot the second one with the `-loadblock` arg here.

**Comment by jgarzik** [comment]
end of an era ;p

utACK

File: test/functional/feature_loadblock.py (hunk 4 above)
Comment (fjahr) [reply test/functional/feature_loadblock.py]: I was weighing different options initially and thought using just one node might be more economical. But I did not think about the logs. I am now using a second node.

**Comment by jonasschnelli** [comment]
Concept ACK

File: ci/test/06_script_b.sh [hunk 5]
```diff
@@ -8,6 +8,9 @@ export LC_ALL=C.UTF-8
 
 cd "build/bitcoin-$HOST" || (echo "could not enter distdir build/bitcoin-$HOST"; exit 1)
 
+mkdir -p contrib
+cp -R ../../contrib/ contrib/
```
Comment (MarcoFalke) [inline ci/test/06_script_b.sh]: I think you can't run these with the `travis` user, only with the docker `root` user.

I.e. 
```suggestion
DOCKER_EXEC cp -R ../../contrib/ contrib/
```

Comment (MarcoFalke) [inline ci/test/06_script_b.sh]: Also, if those files are missing because they are not in the distdir, I'd prefer to add them instead of copying them manually in the ci

File: test/functional/feature_loadblock.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            cfg.write("input={}\n".format(block_dir))
+            cfg.write("genesis={}\n".format(genesis_block))
+            cfg.write("hashlist={}\n".format(hashlist_path))
+
+        base_dir = self.config["environment"]["SRCDIR"]
+        linearize_dir = os.path.join(base_dir, "contrib", "linearize")
+
+        self.log.info("Run linearization of block hashes")
+        linearize_hashes_path = os.path.join(linearize_dir, "linearize-hashes.py")
+        os.system(linearize_hashes_path + " " + cfg_path + " > " + hashlist_path)
```
Comment (MarcoFalke) [inline test/functional/feature_loadblock.py]: Shouldn't you check the return code?

Comment (MarcoFalke) [reply test/functional/feature_loadblock.py]: Also, does the ` > ` work on all OS flavors? It seems that appveyor windows is passing, so it seems fine!?

Comment (laanwj) [reply test/functional/feature_loadblock.py]: also, please use `subprocess` instead of `system` to invoke other processes from python

Comment (fjahr) [reply test/functional/feature_loadblock.py]: > Also, does the `>` work on all OS flavors? It seems that appveyor windows is passing, so it seems fine!?

I did a brief check and did not find any issues concerning windows. I am on macOS and it's fine.

Comment (fjahr) [reply test/functional/feature_loadblock.py]: > also, please use `subprocess` instead of `os.system` to invoke other processes from python

done

> Shouldn't you check the return code?

done

File: ci/test/06_script_b.sh (hunk 5 above)
Comment (fjahr) [reply ci/test/06_script_b.sh]: Yes, that seems cleaner, doing that instead now

**Comment by fjahr** [comment]
As the test got much larger than the initial commit and also fixes a separate issue #17019 it is now split into its own PR at #17091

File: src/init.cpp [hunk 7]
```diff
@@ -1713,8 +1699,16 @@ bool AppInitMain(InitInterfaces& interfaces)
...
 
     std::vector<fs::path> vImportFiles;
-    for (const std::string& strFile : gArgs.GetArgs("-loadblock")) {
-        vImportFiles.push_back(strFile);
+
+    if (gArgs.IsArgSet("-loadblock")) {
+        const std::vector<std::string> vLoadBlockArgs = gArgs.GetArgs("-loadblock");
+
+        // Regard the '-noloadblock' argument
+        if (!vLoadBlockArgs.empty()) {
```
Comment (promag) [inline src/init.cpp]: You could remove this check, iterating an empty array is fine.

**Review by promag — COMMENTED** [review]
Concept ACK

File: src/init.cpp (hunk 7 above)
Comment (fjahr) [reply src/init.cpp]: done

**Comment by luke-jr** [comment]
Concept ACK

File: src/init.cpp [hunk 8]
```diff
@@ -1713,8 +1699,14 @@ bool AppInitMain(InitInterfaces& interfaces)
 #endif
 
     std::vector<fs::path> vImportFiles;
-    for (const std::string& strFile : gArgs.GetArgs("-loadblock")) {
-        vImportFiles.push_back(strFile);
+
+    if (gArgs.IsArgSet("-loadblock")) {
+        const std::vector<std::string> vLoadBlockArgs = gArgs.GetArgs("-loadblock");
+
+        // Regard the '-noloadblock' argument
```
Comment (MarcoFalke) [inline src/init.cpp]: what does this comment mean? `GetArgs` takes care of the `-noloadblock` option

Comment (MarcoFalke) [reply src/init.cpp]: Also, why is this code block changed at all?

File: src/init.cpp [hunk 9]
```diff
@@ -374,7 +374,7 @@ void SetupServerArgs()
     gArgs.AddArg("-debuglogfile=<file>", strprintf("Specify location of debug log file. Relative paths will be prefixed by a net-specific datadir location. (-nodebuglogfile to disable; default: %s)", DEFAULT_DEBUGLOGFILE), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
     gArgs.AddArg("-feefilter", strprintf("Tell other nodes to filter invs to us by our mempool min fee (default: %u)", DEFAULT_FEEFILTER), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::OPTIONS);
     gArgs.AddArg("-includeconf=<file>", "Specify additional configuration file, relative to the -datadir path (only useable from configuration file, not command line)", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
-    gArgs.AddArg("-loadblock=<file>", "Imports blocks from external blk000??.dat file on startup", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
+    gArgs.AddArg("-loadblock=<file>", "Imports blocks from external file on startup", ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
Comment (MarcoFalke) [inline src/init.cpp]: why is this removed? I'd rather add a test for importing `blk000.dat` files.

**Review by maflcko — COMMENTED** [review]
some questions

File: src/init.cpp (hunk 9 above)
Comment (mzumsande) [reply src/init.cpp]: @MarcoFalke: I suggested removing that above, because it seemed to imply a non-existing naming convention for the file (might as well load from `bootstrap.dat` or whatever your file is named).

File: src/init.cpp (hunk 8 above)
Comment (fjahr) [reply src/init.cpp]: True, after implementing the feedback this doesn't do anything different aside from using `AbsPathForConfigVal()` which does not seem to be critical here.

File: src/init.cpp (hunk 9 above)
Comment (MarcoFalke) [reply src/init.cpp]: See for example this test:

```diff
diff --git a/test/functional/feature_loadblock.py b/test/functional/feature_loadblock.py
index bf2a4ff61f..a532a83946 100755
--- a/test/functional/feature_loadblock.py
+++ b/test/functional/feature_loadblock.py
@@ -25,7 +25,7 @@ from test_framework.util import assert_equal, wait_until
 class LoadblockTest(BitcoinTestFramework):
     def set_test_params(self):
         self.setup_clean_chain = True
-        self.num_nodes = 2
+        self.num_nodes = 3
 
     def run_test(self):
         self.nodes[1].setnetworkactive(state=False)
@@ -71,13 +71,16 @@ class LoadblockTest(BitcoinTestFramework):
         subprocess.run([sys.executable, linearize_data_file, cfg_file],
                        check=True)
 
-        self.log.info("Restart second, unsynced node with bootstrap file")
+        self.log.info("Restart unsynced nodes with bootstrap file or raw block file")
         self.stop_node(1)
         self.start_node(1, ["-loadblock=" + bootstrap_file])
-        wait_until(lambda: self.nodes[1].getblockcount() == 100)
+        self.stop_node(2)
+        self.start_node(2, ["-loadblock=" + os.path.join(blocks_dir, 'blk00000.dat')])
+        for i in [1, 2]:
+            wait_until(lambda: self.nodes[i].getblockcount() == 100)
 
-        assert_equal(self.nodes[1].getblockchaininfo()['blocks'], 100)
-        assert_equal(self.nodes[0].getbestblockhash(), self.nodes[1].getbestblockhash())
+            assert_equal(self.nodes[i].getblockchaininfo()['blocks'], 100)
+            assert_equal(self.nodes[0].getbestblockhash(), self.nodes[i].getbestblockhash())
 
 
 if __name__ == '__main__':

**Review by maflcko — COMMENTED** [review]
ACK 5453e2ddf5f9c432d59ba64315a239792136526d

**Comment by laanwj** [comment]
nit:
```patch
diff --git a/doc/developer-notes.md b/doc/developer-notes.md
index 1a7ce91ca67d47903a3b97d45944fb2df6ea2fa1..08004778e3e13284bf1de95efd24a62202847bfc 100644
--- a/doc/developer-notes.md
+++ b/doc/developer-notes.md
@@ -384,7 +384,7 @@ Threads
 
 - ThreadScriptCheck : Verifies block scripts.
 
-- ThreadImport : Loads blocks from blk*.dat files or bootstrap.dat.
+- ThreadImport : Loads blocks from `blk*.dat` files or `-loadblock=<file>`.
 
 - StartNode : Starts other threads.
 
```

File: src/init.cpp (hunk 9 above)
Comment (laanwj) [reply src/init.cpp]: Agree with removing this. `blkXXXXX.dat` are the block files and the way to (re)index them is `-reindex`, it has nothing to do with `-loadblock`.

Comment (MarcoFalke) [reply src/init.cpp]: Isn't the point that you can import the block.dat files from another datadir with `-loadblock=blk001.dat -loadblock=blk002.dat -loadblock=...`

Comment (laanwj) [reply src/init.cpp]: I don't think that's the point. In that case it's less hassle to copy over all the blk files and do a `reindex`.

I think the idea behind `-loadblock` is to import the output from linearize; huge consecutive 'block archives'. But I might be wrong. I'm not sure what it's used for nowadays.

Comment (MarcoFalke) [reply src/init.cpp]: Oh, so it is multi-arg to support a split-up bootstrap.dat?

**Comment by fjahr** [comment]
@laanwj good catch, added that.

**Comment by MarcoFalke** [comment]
ok, fine LGTM

**Comment by laanwj** [comment]
ACK 104f7de5934f13b837fcf21f6d6b2559799eabe2
